libkbfs/folder_block_ops.go

```diff
@@ -2551,12 +2566,14 @@ func (fbo *folderBlockOps) SearchForNodes(ctx context.Context,
 		return nodeMap, nil
 	}
 
-	// Start with the root node
 	rootPtr := md.data.Dir.BlockPointer
 	var node Node
 	if cache == fbo.nodeCache {
```

Comment:
So the node cache used by the main part of KBFS is `fbo.nodeCache`.  This basically maps from `BlockPointer`s to `Node`s.  `Node`s are used by the callers of the library, but internally we need to know the series of `BlockPointer`s and file/dir names that make up the `path` of the corresponding file/dir.  `fbo.nodeCache` is long-lived and never invalidated.

As `folderBranchOps` gets informed of new local or remote MD updates, which change the `BlockPointer`s of some subset of the nodes in this TLF, it calls `nodeCache.UpdatePointer` for each change.  Then, when a caller passes some old `Node` they have lying around into an FBO call, we can translate it to its current `path` using `fbo.nodeCache`.  Note that on every TLF modification, we are guaranteed that the `BlockPointer` of the root directory will change (because of the merkle-ish tree of content hashes we use to assign `BlockPointer`s).

That's a long way of saying we never "invalidate" `fbo.nodeCache`.  It needs to maintain the absolute latest mappings for the TLF, or else FBO calls won't see up-to-date data.  The tension in `SearchForNodes` comes from the fact that we are trying to discover the `BlockPointer`s of certain files _at a specific point in the MD history_, which is not necessarily the same as the most-recently-seen MD update.  Specifically, for the `TlfEditHistory` we are processing a specific range of MDs, but `folderBranchOps` may have heard about a newer one before, or during, `TlfEditHistory` started processing.  That means `fbo.nodeCache` may have been updated to reflect the newest `BlockPointer`s, and is no longer correct as a cache for our search for the data at the old point in time.

So, we basically have two options:
1. Figure out if `fbo.nodeCache` still represents the state of the system at the point of time we care about.  That's what we're doing here by checking that the root pointer for the MD update we care about is still in the cache; if there had been an update in the meantime, that root pointer would have changed.
2. Never search using `fbo.nodeCache`, and always pass in a clean cache.  This would simplify the code, but having a populated node cache allows us to search a bit faster; plus the nodes we create may be useful to other parts of the system later on.

So, that's the background here, and why this code special cases `fbo.nodeCache`.  I'm not sure if that addresses your concern, because I don't know if I fully get what you mean when you say "invalidate `fbo.nodeCache`".